SI-9967 External linking in scaladoc (continued from PR #5461)

haveo opened the pull request:
https://issues.scala-lang.org/browse/SI-9967

Trying to get #5461 over the finish line.

haveo commented:
I signed the CLA

szeiger commented on src/scaladoc/scala/tools/nsc/doc/Settings.scala:
```diff
@@ -253,7 +253,7 @@ class Settings(error: String => Unit, val printMsg: String => Unit = println(_))
   }
 
-  def appendIndex(url: String): String = url.stripSuffix("index.html").stripSuffix("/") + "/index.html"
+  def appendIndex(url: String): String = url.stripSuffix("index.html").stripSuffix("/") + "/"
```
What about `file:` URLs? You can't rely on a rewrite to `index.html`. Linking to the parent dir is a useful option but it should not be the only one and not the default.

haveo replied:
This is probably more of a naming problem as `appendIndex` is only used in `extUrlMapping` which is only used to get the base URL to which the file name is appended. This means that we shouldn't be linking to dirs directly.

szeiger commented on src/scaladoc/scala/tools/nsc/doc/html/HtmlPage.scala:
```diff
@@ -145,8 +145,12 @@ abstract class HtmlPage extends Page { thisPage =>
       <span class="extype" name={ tooltip }>{ inlineToHtml(text) }</span>
-    case LinkToExternal(name, url) =>
-      <a href={ url } class="extype" target="_top">{ inlineToHtml(text) }</a>
+    case LinkToExternalTpl(name, baseUrl, dtpl: TemplateEntity) =>
```
Why does `LinkToExternalTpl` have a `Tpl` type parameter when there is only a single handled case for `TemplateEntity` here? Is the type parameter necessary? If so, can it at least have `TemplateEntity` as an upper bound?

haveo replied:
This is indeed weird. As far as I understand this template is always a `NoDocTemplate` so I'll try to make that explicit. This way the pattern matching should be exhaustive.

szeiger commented on src/scaladoc/scala/tools/nsc/doc/html/HtmlPage.scala:
```diff
@@ -145,8 +145,12 @@ abstract class HtmlPage extends Page { thisPage =>
-    case LinkToExternal(name, url) =>
-      <a href={ url } class="extype" target="_top">{ inlineToHtml(text) }</a>
+    case LinkToExternalTpl(name, baseUrl, dtpl: TemplateEntity) =>
+      def url = Page.makeUrl(baseUrl, Page.templateToPath(dtpl)) + name
```
I don't see any url-encoding. `name` is produced by `MemberLookupBase.externalSignature` without any encoding. `Page.templatePath` uses `NameTransformer.encode`, so the result is *probably* safe but this needs to be double-checked (or just run through the encoder anyway).

haveo replied:
Internal links seem to be handled the same way (append `#` then the signature without any encoding). I'd also be wary of encoding it as a URL as the forbidden characters are not the same. What would you encode it with?

szeiger replied:
Quoting too much won't hurt. <https://docs.oracle.com/javase/8/docs/api/java/net/URI.html#URI-java.lang.String-java.lang.String-java.lang.String-java.lang.String-java.lang.String-> knows how to do it correctly.

szeiger commented on src/scaladoc/scala/tools/nsc/doc/html/Page.scala:
```diff
@@ -13,6 +13,58 @@ import scala.reflect.NameTransformer
+    file :: downPacks(pack)
+  }
+
+  def makeUrl(baseUrl : String, path : List[String]) : String = baseUrl.stripSuffix("/") + "/" + path.reverse.mkString("/")
```
I recommend doing all URI-related computations strongly typed with `java.net.URI` instead of stringly typed to ensure that paths are concatenated correctly and to make it clear what has been encoded and what has not.

I did not look at the rest of the scaladoc code to see how pervasive this kind of URI handling is. If it is already done with string manipulation in many cases a rewrite would best be left for another PR.

haveo replied:
I didn't look at the whole code of scaladoc but `templateToPath` has simply been moved from `Page` to its object companion so I don't have any reason to think it would be much different. A quick `grep` did not seem to find much of any use of `java.net` in scaladoc. I'd rather leave this cleanup to another PR.

szeiger commented on src/scaladoc/scala/tools/nsc/doc/html/Page.scala:
```diff
@@ -13,6 +13,58 @@ import scala.reflect.NameTransformer
+
+  /** A relative link from a page's path to some destination path.
+    * @param destPath The path that the link will point to. */
+  def relativeLinkTo(thisPagePath: List[String], destPath: List[String]): String = {
```
Case in point: With `java.net.URI` this would simply be `thisPagePath.relativize(destPath)`

haveo commented:
@szeiger are there still some changes you need me to make?

szeiger commented:
The changes look good but the commits should be squashed. (I'd generally prefer to keep separate commits from different authors but since the original commit doesn't even compile we shouldn't have it in the history.)

Commit:
Fix externally linked scaladoc in 2.12

Based on work from Jason HiDef Green <jgreen@zynga.com>

Generate NoDocTemplates for external symbols.
Check against isPackage flag rather than Package type (only implemented on DocTemplate) when building URLs

Break out template to path to url methods into companion of Page
Update SI-191 and t8557 tests

haveo commented:
/rebuild

SethTisue commented:
@szeiger is this mergeable for 2.12.3?
